### DISCUSSION TITLE
%titlecase_proper{} - a better %title{} template function with configurability

### DISCUSSION THREAD

**Discussion by trapd00r** [beetbox/beets#4007] [Show and tell]
So it all started [here](https://github.com/beetbox/beets/discussions/3998). 
I wasn't happy with the way the `%title{}` template function worked and tried various things until I came up with a simple plugin that simply used the [titlecase](https://github.com/ppannuto/python-titlecase) library.

It works great with titlecasing everything that should be titlecased (New York Times Manual of Style) while leaving short words alone (a, and, for...).

However, I **still** wasn't completely satisfied though; at times there are words that should neither be titlecased, uppercased or lowercased but rather have a specific casing: for example the acronym 'PhD'. So I whipped up some more code and now the plugin can read **~/.beets_titlecase**, where words are defined one per line, and their case will never be changed. Let's check it out:

                                                                                                                                                               
  ```bash                                                                                                                                                      
  $ cd "Dj Tcp the Reckless │2021│ So Much Drama in the Phd [CDR, MP3]"                                                                                        
  $ for w in TCP DJ PhD; do echo $w; done >> ~/.beets_titlecase                                                                                                
  $ beet import .                                                                                                                                              
  $ beet ls -a 'so much drama' -f \$path                                                                                                                       
  DJ TCP the Reckless │2021│ So Much Drama in the PhD [CDR, MP3]                                                                                               
  ```               
The plugin is available [here](https://github.com/trapd00r/beets-titlecase_proper).
I'm still new to the world of python, so on the todo list is to replace the configuration file with ability to read a list of words to not be touched from `config.yaml`: https://beets.readthedocs.io/en/stable/dev/plugins.html#read-configuration-options

**Comment by sampsyo** [comment]
Super awesome! Let us know if you need help figuring out [the configuration system](https://github.com/beetbox/confuse) to realize that last to-do item.

**Reply by trapd00r** [reply]
**5 hours, 12 minutes and  11 seconds  spent on  titlecase_proper.py**

... And all I got to show for all that time is those few lines of code below. Well, this took _waay_ longer than expected and at one point I was on my way to a 100 sloc before I realized I could simplify things a lot (and then 12 other things broke in the process, repeat while true!). I'm having more fun than I expected learning python this way though.


**beets/config.yaml:**
```
titlecase_proper:
  asis: [
          'PhD',
          'DJ',
          'TCP',
          'YodA',
  ]
```
Renders:
- **DJ Madfist the Invincible TCP Guru │2024│ Got Myself an PhD in the 313 (Deluxe YodA Edition) [MP3]**


It seems to work like it should, but I am unsure about the error handling - is there perhaps a better way to do it in case config.yaml is missing the relevant plugin section?
Other thoughts or input? Much appreciated!

```
from beets         import config
from beets.plugins import BeetsPlugin
from titlecase     import titlecase
import os

# error handling, in case there's no config section for
# titlecase_proper, we should just titlecase as normal
words_to_use_asis = None
try:
  words_to_use_asis = config['titlecase_proper']['asis'].get()
except:
  words_to_use_asis = None

class titlecase_proper(BeetsPlugin):
    def __init__(self):
      super(titlecase_proper, self).__init__()
      self.template_funcs['titlecase_proper'] = _titlecase_proper


def custom_casing(word, **kwargs):
    for w in words_to_use_asis:
      # compare words uppercased, since the casing can be whatever in
      # the config.
        if word.upper() == w.upper():
            return w


def _titlecase_proper(text):
    if words_to_use_asis is not None:
        real_words = titlecase(text, callback=custom_casing)
    else:
        real_words = titlecase(text)

    return ''.join(real_words)

```
